caikit/core/task.py

```diff
@@ -0,0 +1,180 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Standard
+from typing import Callable, Dict, List, Set, Type, Union
+
+# First Party
+from alog import alog
+
+# Local
+from caikit.core.data_model import DataStream
+from caikit.core.data_model.base import DataBase
+from caikit.core.toolkit.errors import error_handler
+
+log = alog.use_channel("TASK_BASE")
+error = error_handler.get(log)
+
+ProtoableInputTypes = Type[Union[int, float, str, bytes, bool, DataBase]]
+ValidInputTypes = Union[
+    ProtoableInputTypes, List[ProtoableInputTypes], DataStream[ProtoableInputTypes]
+]
+
+
+class TaskGroupBase:
+    @classmethod
+    def validate_task_inputs(cls) -> bool:
+        pass
+
+    @classmethod
+    def get_input_type_set(cls) -> Set[ProtoableInputTypes]:
+        raise NotImplementedError("This is implemented by the @taskgroup decorator!")
+
+
+class TaskBase:
+    @classmethod
+    def validate_run_signature(cls) -> bool:
+        # TODO: implement
+        pass
+
+    @classmethod
+    def get_required_inputs(cls) -> Dict[str, ValidInputTypes]:
```

Comment:
Even though these methods will always be overwritten, they should have docstrings so that `help(TaskBase.get_required_inputs)` works and gives useful info about what the method does in the abstract.